Version check Yuhang Liao

ComarPers922 opened the pull request:
UPDATE FROM YUHANG LIAO

1. Version Check feature is added. Type "info" to get current version information.

2. Modify some error messages.

The previous pull request has been closed and invalidated.

ComarPers922 committed:
Version Check From Yuhang Liao

ComarPers922 committed:
Add some frinedly output and modify some comments

ComarPers922 committed:
Friendly output

ComarPers922 committed:
Modify some comments and output

ComarPers922 committed:
ServerInfo Modified

ComarPers922 committed:
1. Version Check feature is added. Type "info" to get current version information.

2. Modify some error messages.

ComarPers922 committed:
UPDATE FROM YUHANG LIAO

lepidopterane-atsmith commented:
This Code Review is going to be a lot like the last one. Of course reach out to me on Slack! **I only have a few bolded questions which I expect you to answer on our channel.**

lepidopterane-atsmith commented:
First, don’t forget, you can and should learn how to do the thing Alex taught you how to do via git where you only upload the changed/brand new files.

lepidopterane-atsmith commented:
_readme.md_
- It’s good to see you document updates in your readme!
- But the error that was always in your readmes is still there
- Plus, “Clarified error messages” is a lot less vague than “Modify some error messages”

lepidopterane-atsmith commented:
_Chat.java_
- Your error message looks good here!
- The definition you use in lines 113 and 120 sounds vague and unsure, so you should just say “Get session information.”

lepidopterane-atsmith commented:
_Context.java_
I understand what the code was trying to accomplish here, but it does not need an if-else to govern it. View.getInfo should always be called because it will return what needs to be returned, so if it returns a null it has printed the relevant error messages to explain that instance.

lepidopterane-atsmith commented:
_View.java_
- Line 153 is nice and functional.
- I’m not sure about line 157, though. **Can you explain what was the reasoning behind that line? If it helps, maybe you should create a test file for tripping that error (If you do, include it in your next pull!)**

lepidopterane-atsmith commented:
_ServerInfo.java_
Last pull, I explained why a try/catch statement was unnecessary in a ServerInfo constructor. This also goes for try/catch/finally.

aslo commented on README.md:
```diff
@@ -1,6 +1,11 @@
 
 # CODEU CHAT SERVER | README
 
+## UPDATE FROM YUHANG LIAO
```
I don't think this is the appropriate place to add this information. This README is focused on how to build and run the application, and should be kept as concise as possible. Information about new features being added belong in a Pull Request message, or commit message instead. 

I recommend deleting this `UPDATE FROM YUHANG LIAO` section of the README.

aslo replied:
One other big issue - you've changed the file permissions of many files in this pull request. Notice that at the top, many files display a change like this:

![screen shot 2017-05-29 at 2 56 22 pm](https://cloud.githubusercontent.com/assets/1039411/26559668/03f9fde4-447f-11e7-8cfa-8e09f8edb2af.png)

That means these files have been changed to an "executable" status, which is incorrect - they are not executable files, and this can be confusing to other developers in the future. More importantly, a pull request should only change the files relevant to the feature being worked on. Before merging this, I strongly recommend that you change these files back to the default permissions. Assuming you're on Linux or Mac, for each of these files, you should run `chmod 0644 <file_name_here>` to revert them back to the correct status.

[This stackoverflow answer](https://stackoverflow.com/a/8347325/5053647) has more info about file permissions in case you're interested.

aslo commented on README.md:
````diff
@@ -34,15 +39,15 @@ JAVA&nbsp;7 when working with this project.
   1. To test the project:
        ```
-       $ python build.py run codeu.chat.TestRunner
+       $ python build.py codeu.chat.TestRunner
````
This might not have been intentional on your part, but you should change this line back to how it was originally.

lepidopterane-atsmith replied:
^^ What Alex said

aslo commented on README.md:
````diff
@@ -34,15 +39,15 @@ JAVA&nbsp;7 when working with this project.
        ```
        $ python build.py run codeu.chat.ServerMain <team_id> <team_secret> <port> <persistent-dir>
-       $ python build.py run codeu.chat.ClientMain "<host>@<port>"
+       $ python build.py run codeu.chat.Clientmain "<host>@<port>"
````
Same as above, this might not have been intentional on your part, but you should change this line back to how it was originally.

lepidopterane-atsmith replied:
^^ What Alex said

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -107,11 +109,30 @@ public void invoke(Scanner args) {
+          if (info == null) {
+            // Communicate error to user - the server did not send us a valid
+            // info object.
+            System.out.println("Error: ServerInfo cannot be read. Line 125, Chat.java");
```
Generally it's not a good idea to reference file names or line numbers in an error message like this, because in the future, someone might move the code around or rename the file, and forget to update this line. A better way to communicate this information to the user is to use an exception and a stack trace, which will always be up to date with the files and line numbers. 

If you want to make sure this information is included with the error message, try replacing this line with something like:
```
new IllegalStateException("Error: ServerInfo cannot be read.").printStackTrace();
```

aslo commented on src/codeu/chat/client/core/View.java:
```diff
@@ -136,4 +137,26 @@ public View(ConnectionSource source) {
+      } else {
+        // Communicate this error - the server did not respond with the type of
+        // response we expected.
+        System.out.println("Error: The server did not respond with the type of response we expected. Line 146-149, View.java.");
```
Same issue here, remove file names and line numbers.

aslo commented on src/codeu/chat/client/core/View.java:
```diff
@@ -136,4 +137,26 @@ public View(ConnectionSource source) {
+      }
+    } catch (Exception ex) {
+      // Communicate this error - something went wrong with the connection.
+      System.out.println("Fatal error: Unexpected error occurs. Line 146-149, View.java" + ex.getMessage());
```
Same issue here, remove file names and line numbers.

aslo commented on src/codeu/chat/client/core/Context.java:
```diff
@@ -47,4 +48,19 @@ public UserContext create(String name) {
+    }
+    else
+    {
+      System.out.println("Fatal Error: The info cannot be parsed.Line 58, Context.java");
```
Same issue as above - you shouldn't reference file names or line numbers like this in error messages. Consider changing this line. However, if you follow my comment to remove the class casting in this method, this line can just be deleted anyway.

aslo commented on src/codeu/chat/client/core/Context.java:
```diff
@@ -47,4 +48,19 @@ public UserContext create(String name) {
+  public ServerInfo getInfo() {
+    if(view instanceof View)
+    {
+      return ((View)view).getInfo();
```
Generally, you should avoid class casting like this if possible, since it has the opportunity to introduce runtime exceptions into the code. You should just change the declaration of this variable to look like `private final View view;` on line 29. Then this entire method can just be one line: 
```
return view.getInfo();
```

aslo reviewed (changes requested):
Nice job! I left you some comments and suggestions to fix before merging this pull request. Let me know if you have any questions!

ComarPers922 committed:
1.Change some error messages.
2.Revert permissions.

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -126,7 +127,7 @@ public void invoke(Scanner args) {
             // Communicate error to user - the server did not send us a valid
             // info object.
-            System.out.println("Error: ServerInfo cannot be read. Line 125, Chat.java");
+            new IOException().printStackTrace("ERROR: ServerInfo cannot be read.");
```
printStackTrace doesn't take string parameters. Also what would be a stack trace for a new IOException?

lepidopterane-atsmith commented on src/codeu/chat/client/core/View.java:
```diff
@@ -150,11 +152,11 @@ public ServerInfo getInfo() {
     } catch (Exception ex) {
       // Communicate this error - something went wrong with the connection.
-      System.out.println("Fatal error: Unexpected error occurs. Line 146-149, View.java" + ex.getMessage());
+      new UnexpectedException().printStackTrace("FATAL ERROR: Unexpected error occurs.");
```
1. Why is this a fatal error?
2. An unexpected error could happen that would have nothing to do with the connection, but the one that got us here did. This line does not express that.
3. This line has the same error w/printStackTrace syntax as the previous line I corrected. It should be:
System.out.println("insert error message here");
ex.printStackTrace();

ComarPers922 replied:
I really don't know how to explain an unexpected error...

lepidopterane-atsmith commented on src/codeu/chat/common/ServerInfo.java:
```diff
@@ -12,7 +12,7 @@ public ServerInfo(){
         temp = Uuid.parse(SERVER_VERSION);
       } catch (IOException e) {
-          System.out.println("Fatal Error: Cannot parse the server infomation. Line 13, ServerInfo.java");
+          e.printStackTrace("Fatal Error: Cannot parse the server infomation. All variables in this instance will be null this time.");
```
1. What makes this error fatal?
2. Same syntax problem w/printStackTrace, different lines.
3. Second error sentence is unnecessary because nothing guaranteeable happens in the program after this message and its syntax is printed.

ComarPers922 replied:
I will fix the parameter problem later.

lepidopterane-atsmith reviewed (changes requested):
You're getting better at this! I have a few questions, but really you just need to get a feel for how printStackTrace() works.

ComarPers922 commented on src/codeu/chat/common/ServerInfo.java:
```diff
@@ -12,7 +12,7 @@ public ServerInfo(){
         temp = Uuid.parse(SERVER_VERSION);
       } catch (IOException e) {
-          System.out.println("Fatal Error: Cannot parse the server infomation. Line 13, ServerInfo.java");
+          e.printStackTrace("Fatal Error: Cannot parse the server infomation. All variables in this instance will be null this time.");
```
Please notice I just use one line to print the error, which is printStackTrace.

ComarPers922 commented on src/codeu/chat/client/core/View.java:
```diff
@@ -150,11 +152,11 @@ public ServerInfo getInfo() {
     } catch (Exception ex) {
       // Communicate this error - something went wrong with the connection.
-      System.out.println("Fatal error: Unexpected error occurs. Line 146-149, View.java" + ex.getMessage());
+      new UnexpectedException().printStackTrace("FATAL ERROR: Unexpected error occurs.");
```
Please notice I just use one line to print the error, which is printStackTrace.

ComarPers922 committed:
1.Fix the parameter problem.
2.Change "Scanner" to "List<Sting>"
3.I need to tell users an unexpected exception occurs. I can't tell users what really happens, because it is unexpected, which means unknown.

ComarPers922 committed:
Merge branch 'master' into Version-Check-Yuhang-Liao

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -14,20 +14,25 @@
+
 import java.util.ArrayList;
 import java.util.List;
 import java.io.IOException;
```
This is just me being nitpicky but you should delete this line since it is a copy of line 17

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -120,11 +126,30 @@ public void invoke(List<String> args) {
         System.out.println("  u-sign-in <name>");
         System.out.println("    Sign in as the user with the given name.");
+        System.out.println("  info");
+        System.out.println("    Get some information from server; it should be version info currently.");
```
"some information" is vague, and don't say anything after the semicolon. One way to do this is to replace that text with "Get session information."

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -120,11 +126,30 @@ public void invoke(List<String> args) {
-
+    // info (Server info)
+    //
+    // Get some infomation from server; it should be version info currently
```
"some information" is vague, and don't say anything after the semicolon. One way to do this is to replace that text with "Get the follwing session information: Version Info." Don't worry, that list will grow when Vicente has to merge his stuff with yours.

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -120,11 +126,30 @@ public void invoke(List<String> args) {
+          if (info == null) {
+            // Communicate error to user - the server did not send us a valid
+            // info object.
+            new IOException("ERROR: ServerInfo cannot be read.").printStackTrace();
```
That format only works with try/catch(Exception e) scenarios; a simple System.out.println("ERROR: ServerInfo cannot be read."); would do. Also, when I ran this, I got a different error because there are no IOException constructors that just take a String.

lepidopterane-atsmith commented on src/codeu/chat/client/core/View.java:
```diff
@@ -14,6 +14,8 @@
 package codeu.chat.client.core;
 
+import java.io.IOException;
+import java.rmi.UnexpectedException;
```
We don't need this and I'll tell you why

lepidopterane-atsmith commented on src/codeu/chat/client/core/View.java:
```diff
@@ -136,4 +139,26 @@ public View(ConnectionSource source) {
+      } else {
+        // Communicate this error - the server did not respond with the type of
+        // response we expected.
+        new IOException("ERROR: The server did not respond with the type of response we expected.").printStackTrace();
```
That format only works with try/catch(Exception e) scenarios, and can only be implemented in the catch brackets. A simple System.out.println("ERROR: The server did not respond with the type of response we expected."); would do. Also, when I ran this, I got a different error because there are no IOException constructors that just take a String.

lepidopterane-atsmith commented on src/codeu/chat/client/core/View.java:
```diff
@@ -136,4 +139,26 @@ public View(ConnectionSource source) {
+      }
+    } catch (Exception ex) {
+      // Communicate this error - something went wrong with the connection.
+      new UnexpectedException("FATAL ERROR: Unexpected error occurs, something went wrong while client was trying to connect to server.").printStackTrace();
```
I understand what sort of format you are trying to do here, but here's the format in which it should be done:
catch (Exception ex) {
System.out.println("<insert error message here>");
ex.getStackTrace();

Also, it's important to note that this is not a fatal or unexpected error, or else the comments would have told us so.

lepidopterane-atsmith commented on src/codeu/chat/common/ServerInfo.java:
```diff
@@ -0,0 +1,27 @@
+  public ServerInfo(){
+      Uuid temp = null;
+      try {
+        temp = Uuid.parse(SERVER_VERSION);
```
SERVER_VERSION is final and static from the time it's initialized in the class, so it's a constant value that will never change. I should also point out that any time a constant like this is initialized, they usually have NAMES_LIKE_THIS . Because of these two things, a try/catch/finally is unnecessary in the constructor because if a value never changes to something it shouldn't be and is always initialized when it needs to be, it will be in a constant, non-erroneous state 24/7.

aslo replied:
Just wanted to chime in here. A try/catch block is actually necessary here because the java compiler requires it. IOExceptions are a class of exceptions that we need to catch in our code, otherwise it just won't compile at all :)

lepidopterane-atsmith commented on src/codeu/chat/client/commandline/Panel.java:
```diff
@@ -49,12 +49,10 @@ public void register(String commandName, Command command) {
+  public boolean handleCommand(String commandName, List<String> args) {
+  final Command command = commands.get(commandName);
+  if (command != null) {
+    command.invoke(args);
```
Good job with averting merge conflicts!

lepidopterane-atsmith reviewed (changes requested):
Okay, because GitHub is special I reviewed your last two pulls (merging and previous changes because they are both significant). You did well with merging, but there are still some areas you need to work on. **Also, you should tell me about the last time you compiled and ran your commit.**

ComarPers922 committed:
Correct the README

ComarPers922 committed:
Correct the output

ComarPers922 committed:
Merge branch 'Version-Check-Yuhang-Liao' of github.com:vgomezASC/CodeU-Summer-2017 into Version-Check-Yuhang-Liao

aslo reviewed (changes requested):
Hi @ComarPers922, this is looking really great! Thanks for addressing the feedback from my last review. I think this is almost good to merge! 

It looks like there are a few files whose permissions you still need to change back: 
* .travis.yml
* .gitignore
* README.md

Once you fix that, I'll be happy to approve this!

ComarPers922 committed:
Ready to merge

lepidopterane-atsmith commented:
Unless Alex or anyone else has something to say here, LGTM.
